packages/shared/src/hooks/streaks/useStreakMilestone.ts

```diff
@@ -0,0 +1,55 @@
+import { useContext, useEffect } from 'react';
+import { useLazyModal } from '../useLazyModal';
+import { LazyModal } from '../../components/modals/common/types';
+import AlertContext from '../../contexts/AlertContext';
+import { useSettingsContext } from '../../contexts/SettingsContext';
+import { useReadingStreak } from './useReadingStreak';
+import { StreakModalProps } from '../../components/modals/streaks/common';
+
+export const useStreakMilestone = (): void => {
+  const { loadedSettings, optOutWeeklyGoal } = useSettingsContext();
+  const { streak, isLoading, isEnabled } = useReadingStreak();
+
+  const { loadedAlerts, alerts, updateAlerts } = useContext(AlertContext);
+  const { openModal, closeModal } = useLazyModal();
+
+  const modalType =
+    streak?.total === 1 ? LazyModal.FirstStreak : LazyModal.NewStreak;
+
+  useEffect(() => {
+    if (
+      !loadedAlerts ||
+      alerts?.showStreakMilestone !== true ||
+      isLoading ||
+      !isEnabled ||
+      !loadedSettings ||
+      optOutWeeklyGoal ||
+      !streak?.current
+    ) {
+      return;
+    }
+
+    openModal({
+      type: modalType,
+      props: {
+        currentStreak: streak?.current,
+        onRequestClose: () => {
+          updateAlerts({ showStreakMilestone: false });
+          closeModal();
+        },
+      } as StreakModalProps,
```

Comment:
Because the dynamic modalType Ts issue?

Reply:
We have a generic available on openModal right, or does that also not work?

Reply:
This is probably because the modal has not yet been introduced to the list of modals at `/components/modals/common.tsx`

Reply:
Both of the modal types were introduced in `components/modals/common.tsx`. This is the error I was getting from TS compiler:
```
Object literal may only specify known properties, and 'onRequestClose' does not exist in type 'LazyModalComponentType<LazyModal.SquadMember> | LazyModalComponentType<LazyModal.SquadTour> | LazyModalComponentType<LazyModal.UpvotedPopup> | ... 13 more ... | LazyModalComponentType<...>'.ts(2353)
common.tsx(161, 9): The expected type comes from property 'props' which is declared here on type 'LazyModalType<LazyModal>'
```

Reply:
It's because we omit them.
```
type LazyModalComponentType<K extends LazyModal> = Omit<
  GetComponentProps<ModalsType[K]>,
  'isOpen' | 'onRequestClose'
>;
```

I think the idea was that these should never be interfered with as they do a lot of magic stuff.
Some alternatives is to pass a onAfterClose *(some modal has this already)

Reply:
TIL, thanks for the input 👍